File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        return solve(root, targetSum);
    }
private:
    void backtrack(TreeNode* root, int& targetSum, vector<int>& partialAns, vector<vector<int>>& ans) {
        if(root == nullptr)
            return;
        
        targetSum -= root -> val;
        partialAns.push_back(root -> val);
        
        if(root -> left == nullptr && root -> right == nullptr) {
            if(targetSum == 0)
                ans.push_back(partialAns);
            
            targetSum += root -> val;
            partialAns.pop_back();
            return;
        }
        
        
        if(root -> left) backtrack(root -> left, targetSum, partialAns, ans);
        if(root -> right) backtrack(root -> right, targetSum, partialAns, ans);
        
        partialAns.pop_back();
        targetSum += root -> val;
        return;
    }
    
    
    vector<vector<int>> solve(TreeNode* root, int targetSum) {
        vector<int> partialAns;
        vector<vector<int>> ans;
        
        backtrack(root, targetSum, partialAns, ans);
        
        return ans;
    }
}; 

File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def backtrack(self, root, targetSum):
        
        # include the current root value
        targetSum -= root.val
        self.partialAns.append(root.val)
        
        # if we are at leaf node
        if root.left == None and root.right == None:
            
            # if path sum is equal to targetSum
            if targetSum == 0:
                self.ans.append(self.partialAns[:])
            
            # now pop the leaf element
            # this operation is done here because we are returning from here
            # and control is not reaching line - 37, 38 for the same
            targetSum += root.val
            self.partialAns.pop()
            return
        
        # if left subtree is not none
        if root.left != None:
            self.backtrack(root.left, targetSum)
        
        # if right subtree is not none
        if root.right != None:
            self.backtrack(root.right, targetSum)
        
        # backtracking step
        targetSum += root.val
        self.partialAns.pop()
        
        
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root == None:
            return []
        
        self.partialAns = []
        self.ans = []
        
        self.backtrack(root, targetSum)
        
        return self.ans